### ISSUE TITLE
Overlap with Agents.jl

### ISSUE THREAD

**Issue by Datseris** [cesaraustralia/DynamicGrids.jl#10]
Hey there,

I just found this gem, and the other gem Dispersal.jl, after your post on [Discourse](https://discourse.julialang.org/t/how-can-we-create-a-leaner-ecosystem-for-julia/32904/30), and I was totally unaware of it until then (yet one more benefit of orgs.).

Here I just want to discuss that there is certainly overlap between this one and Agents.jl, as we both do cellular automata as well as agent based models in some sense. What might be useful for both of us is to compare the implementations and see which ones are better from the perspective of

* performance
* api 
* extandability (if this is even possible with CA)
* visualization (its already pretty clear you have a very strong case here with these terminal vizs)

In general in JuliaDynamics we always look out for the best implementations. I am tagging @kavir1698 , the main dev of Agents.jl, certainly this will be interesting for him as well. How Agents.jl is structured and operates is summarized neatly in the short [Tutorial page](https://juliadynamics.github.io/Agents.jl/dev/tutorial/). 

Do you think you could maybe summarize the way DynamicGrids.jl works from a developers perspective, so that it is easier to compare?

**Comment by rafaqz** [comment]
Good to hear from you! I have been thinking Agents.jl could build on DynamicGrids.jl at some stage. It's made to be an engine to build other packages on - the `Life` rule is just good to have for an example and for testing - but it has a lot of neighborhood options.

I put together this developer summary, a lot of this has been made up as we go so its not so easy to describe in clear language.

So, DynamicGrids is a simulation engine that runs rules over a grid of cells (ie an array), for some number of timesteps/frames. It applies performance optimisations to running the rules while keeping everything as generic and flexible as possible. It's only 2d currently, but it could do 1 and 3 as well, I just don't need them yet. 

Rules are just structs with an `applyrule` or `applyrule!` method. There are different classes of rules that allow different operations, and have different performance optimisations to match their limitations. `CellRule`s operate on one cell and return a new cell value. So they can be chained without read/write. `NeighborhoodRule`s operate on a neighborhood and return a new cell value. `PartialRules` do not write a return value, and instead write to arbitrary points in the grid manually - for random dispersal jumps etc. Rules have all the simulation data available to them if they need it, but generally they don't and should avoid digging into it to maintain modularity.

`Rule`s can be combined arbitrarily in `Ruleset`s ie `Ruleset(rule1, rule2, rule3)` and applied in sequence over each frame. Interactions can also occur between multiple grids - and these can also be chained arbitrarily and attached to the correct grids using named tuples and keys.

Each frames of the simulation is either visualised live in `Output`s or stored in an array or both, or whatever you want to write custom.

I'll also try and answer the dot points:

**- performance**

DynamicGrids has had some serious performance optimisation work. We use it in Optim.jl with replicates and end up running billions of 500*500 frames, so we've had to optimise really hard.
- It does buffer swapping at each timestep
- It adds padding to avoid neighborhood bounds checking
- It doesn't run empty (zero) neighborhoods, which can give an order of magnitude speedup depending on density. It tracks blocks and monitors all array writes to achieve this. 
- It copies grid data to local neighbourhood buffer efficiently, where you can use BLAS on it.
- Some rule types can be chained, and compiled together into single broadcasts skipping array read/write.

*- api*
Its meant to be very generic and flexible. I'm not sure if its totally there yet but close. It's easy to add custom rules and interaction rules. The [Gtk](https://github.com/cesaraustralia/DynamicGridsGtk.jl/blob/master/src/DynamicGridsGtk.jl) interface is only 66 lines, so also easy to add your own custom interface too.

*-expandability*
- you can chain as many custom models together as you want in any order - they are applied recursively. 
- You can use as many interacting grids as you like - in the gif in the readme, some of those grids are feeding back on each other using Interaction rules.
- You can add your own abstract Rule types and add new optimisation methods for running them. 
- It's all generic multiple dispatch so there's basically nothing you can't customise, and I'm trying to make that easy as much as possible.

*-visualisation*
- Terminal, Gtk, and web browser, electron, atom and jupyter are available. 
- The web interface has live Interact.jl controls even for arbitrary custom models (using Flatten.jl). 
- It does multi image layouts and RBG band layers with a pretty nice syntax. 
- The dispersal simulation (top left) on its own runs 25 replicates at 10fps overlayed live on my laptop - so it's pretty well optimised for real-time visualisation.

It was really only released to make working on Dispersal.jl easier, for Documenter.jl and Travis. But it's mostly useable already, and I've been making fewer and fewer breaking changes. Some feedback on naming conventions would be great before they get too locked in...

**Comment by Datseris** [comment]
Thanks for the reply!

The performance and visualization parts seems very strong! If I have understood everything, this library seems to be very much suited for regular grid spaces. Maybe we can implement one more `Space` type, that uses internally DynamicGrids.jl, or this engine, for very large simulations of CA-like systems. But we need a simple working example first :D 

I am trying to read the documentation (the Examples), but I must admit, not having an example "from 0 to hero", that does everything a user would do makes the API less transparent. Do you think you can consider starting the docs with an example that uses _nothing_ already implemented? Calling `Life()` is not a great first point, as it requires me as a user to first learn `Life`, before learning DynamicGrids.jl. For example, I'd like to implement the `forest_fire` model from Agents.jl: https://juliadynamics.github.io/Agents.jl/dev/examples/forest_fire/ .

> Rules can be combined arbitrarily in Rulesets ie Ruleset(rule1, rule2, rule3) and applied in sequence over each frame. Interactions can also occur between multiple grids - and these can also be chained arbitrarily and attached to the correct grids using named tuples and keys.

Hm, this is an interesting idea that we haven't considered in Agents.jl. So far our "rules" are given always in the form of one core function that does a "step!". Compos-ability from simpler parts seems nice, but I also have to say that it is much harder to do in more complex, non game-of-life agent models.

**Comment by Datseris** [comment]
One more point, but important distinction with `GridSpace` from Agents.jl: Is it possible is it in DynamicGrids.jl to have multiple "agents" in one cell of the grid?

**Comment by rafaqz** [comment]
Look at the Dispersal.jl examples, there is a walk through there. I don't really have time to make examples to the same quality for DynamicGrids.jl at present, as it's not really user-facing and I have too many packages to prioritise first. Reading Dispersal.jl code and tests might be a good overview of how it works.

Grids can hold anything you want! It's just an array. The value of the cell will be passed into the rule, whatever it is. But a grid of (variable length?) vectors of individuals isn't something I've tried or optimised at all - it should also work fine with a little work. You will have to write a frame processor for visualising the grid in Gtk/Web or whatever - converting the cell to an `RGB24` somehow or other.

We mostly model populations of identical individuals (eg insects) so the cells are just floats or ints, but size-structured populations in static arrays would work fine as well.

**Comment by rafaqz** [comment]
One more thing, Life() is actually really simple (except the metadata markup lol). This is all the code:

```julia
@description @limits @flattenable @default_kw struct Life{R,N,B,S} <: NeighborhoodRule{R}
    neighborhood::N | RadialNeighborhood{1}() | false | nothing | "Any Neighborhood"
    b::B | (3, 3) | true | (0, 8) | "Array, Tuple or Iterable of integers to match neighbors when cell is empty"
    s::S | (2, 3) | true | (0, 8) | "Array, Tuple or Iterable of integers to match neighbors cell is full"
end
Life(neighborhood::N, b::B, s::S) where {N,B,S} = 
    Life{radius(neighborhood),N,B,S}(neighborhood, b, s) 

applyrule(rule::Life, data, state, index, buf) = begin
    # Sum neighborhood
    cc = neighbors(rule.neighborhood, rule, buf, state)
    # Determine next state based on current state and neighborhood total
    counts = (rule.b, rule.s)[state+1]
    for i = 1:length(counts)
        if counts[i] == cc
            return oneunit(state)
        end
    end
    return zero(state)
end
```

I may as well paste that into the example.

**Comment by Datseris** [comment]
> I may as well paste that into the example.

I need to tell you that this:
```julia
@description @limits @flattenable @default_kw struct Life{R,N,B,S} <: NeighborhoodRule{R}
    neighborhood::N | RadialNeighborhood{1}() | false | nothing | "Any Neighborhood"
    b::B | (3, 3) | true | (0, 8) | "Array, Tuple or Iterable of integers to match neighbors when cell is empty"
    s::S | (2, 3) | true | (0, 8) | "Array, Tuple or Iterable of integers to match neighbors cell is full"
end
```
is **astonishingly** complicated Julia code... You might want to keep only basic Julia code in introductory examples.

> `applyrule(rule::Life, data, state, index, buf) = begin`

It is more reasonable to turn this into a `function` definition than a begin block.

(these are just documentation advices, take them or leave them)

**Comment by rafaqz** [comment]
Oh I know lol that's why I mentioned it in brackets. It's not easy getting auto-generated interfaces for arbitrary models. That of course would not go in the docs, but I'm not going to pretend in this issue that the code is simpler than what it is.

Also I hate function definitions because I hate switching when you drop to one line, and most of my methods are 1-3 lines. The rest of Julia will eventually come round ;)

**Comment by rafaqz** [comment]
Looking at Agents.jl I think maybe these are fundamentally different concepts. Agents.jl loops over agents, DynamicGrids.jl broadcasts over the grid. 

To use tracked agents the grid would be the boolean status, and the rule would have to hold the agents list. You would then have to find the agents for the current cell from the list, which might cancel out the gains from all the other optimisations. It won't scale past small agent lists. 

So - I'm not sure if there is much to share between these packages, besides the visualisation outputs.

**Comment by rafaqz** [comment]
Ok heres a solution for small numbers of agents: each bit represents an agent. So you can track 63 agents in a `UInt64` grid, 127 in Unit128 etc.

 Zero is empty status, non-zero is occupied. Then you loop over the bits for each cell and lookup the corresponding agent. Then use the same neighbourhood for each. It will still hit scaling problems at some point, but much later than actually searching the agents list. And the neighbourhoods will be very fast as they are also being reused.

It might perform pretty well. The repl output will just work if the threshold is set to zero, and the other output processors should be fairly easy to write.

I'll code up a demo if I ever get time.

**Comment by Datseris** [comment]
> Looking at Agents.jl I think maybe these are fundamentally different concepts. Agents.jl loops over agents, DynamicGrids.jl broadcasts over the grid.

This is true, but it is done so because there are occassions where there are multiple agents in one grid point. But it doesn't have to be like this. For example all Cellular Automata, and the forest fire model I've mentioned, are 1-agent-per-cell, which means that the agent and the cell are equivalent and interchange-able. What I imagine as an interface is one that broadcasts the agent's 'status' or value or whatever, to values on the grid and then runs the simulation with DynamicGrids.jl.

**Comment by rafaqz** [comment]
Ok thats interesting. It would definately make sense to use DynamicGrids for those.

**Comment by Datseris** [comment]
Yeap, I think it is definitely worth it for us to provide an interface to DynamicGrids.jl via one additional `Space` type or something along these lines. First step is making `forest_fire` here, to get an idea of the interfacing.

**Comment by rafaqz** [comment]
Ok here you go, with a Gtk output and saving a gif.

```julia
using DynamicGrids, DynamicGridsGtk, ColorSchemes, Colors

# Define the model
const DEAD = 0
const BURNING = 1
const ALIVE = 2

struct ForestFire{R,N,P} <: NeighborhoodRule{R}
    neighborhood::N
    prob_combustion::P
end
ForestFire(; neighborhood=RadialNeighborhood{1}(), prob_combustion=0.0001) =
    ForestFire{DynamicGrids.radius(neighborhood),typeof(neighborhood),typeof(prob_combustion)
              }(neighborhood, prob_combustion)

@inline DynamicGrids.applyrule(rule::ForestFire, data, state::Integer, index, hood_buffer) =
    if state == ALIVE
        if BURNING in hood_buffer
            BURNING
        else
            rand() <= rule.prob_combustion ? BURNING : ALIVE
        end
    else
        DEAD
    end

# Run a simulation
init = fill(ALIVE, 400, 400)
ruleset = Ruleset(ForestFire(); init=init)
output = GtkOutput(init; store=true, fps=10, minval=0, maxval=2,
                   processor=ColorProcessor(scheme=ColorSchemes.brg, zerocolor=RGB24(0.0)))
sim!(output, ruleset; tspan=(1, 50))

# Save a gif
savegif("fire.gif", output)
```

![fire](https://user-images.githubusercontent.com/2534009/71825602-4e2bd080-30f0-11ea-9d49-501369c1d04b.gif)

**Comment by rafaqz** [comment]
Most of the time is taken by visualisation, running it with a non-visualised ArrayOutput:
```julia
output = ArrayOutput(init, 50)
julia> @time sim!(output, ruleset; tspan=(1, 50))
  0.146112 seconds (190 allocations: 2.530 MiB)
```
That's about 350 frames a second at 400*400, on a laptop.

Writing this up also made me realise a few things about design. There should be a method that returns the neighborhood as an iterator so you can count particular values in any neighbourhood shape - like burning neighbors. There are custom neighbourhoods with arbitrary shapes you could use instead of the basic Moor neighborhood here. But they don't work as iterators and instead sum, which doesn't make sense in this case. There is a `mapreduceneighbors` method, but it's a bit clunky.

Also having the radius stored as the type parameter `R` actually doesn't help performance much, making it a regular field would be cleaner.

**Comment by Datseris** [comment]
Seems very good. It only misses the part where new trees can grow in `DEAD` cells, based on some other probability, but probably implementing this is trivial by modifying the `else` branch to not just always return `DEAD` but make some cells alive again.

> That's about 350 frames a second for a 400*400 grid, on a laptop.

Okay, I'll benchmark how much time our implementation takes, with re-growth probability zero, and report back here.

Is the neighborhood here Moore, i.e. 8 neighbors, or 4? By reading `RadialNeighborhood{1}()` I would assume 4, because `1` would mean the radius in this case. Is this correct?

**Comment by rafaqz** [comment]
Ok I missed the regrowth bit. It actually won't work as-is because of the ignore-zeros optimisation. You would need to use `DEAD = 3` and set `maxval=3` in the visualisation.

I will eventually add a keyword arg to turn the optimisation off.

And its 8 neighbours. It's a Von Neuman neighbourhood that has 4 neighbours with a radius 1. DG has that too, but as I said it needs an iterator to make this code generic to accept any neighbourhood.

**Comment by rafaqz** [comment]
Ok here's the fixed version.

```julia
using DynamicGrids, DynamicGridsGtk, ColorSchemes, Colors

const DEAD = 1
const ALIVE = 2
const BURNING = 3

struct ForestFire{R,N,PC,PR} <: NeighborhoodRule{R}
    neighborhood::N
    prob_combustion::PC
    prob_regrowth::PR
end
ForestFire(; neighborhood=RadialNeighborhood{1}(), prob_combustion=0.0001, prob_regrowth=0.01) =
    ForestFire{DynamicGrids.radius(neighborhood),typeof.((neighborhood, prob_combustion, prob_regrowth))...
              }(neighborhood, prob_combustion, prob_regrowth)

@inline DynamicGrids.applyrule(rule::ForestFire, data, state::Integer, index, hood_buffer) =
    if state == ALIVE
        if BURNING in hood_buffer
            BURNING
        else
            rand() <= rule.prob_combustion ? BURNING : ALIVE
        end
    elseif state in BURNING
        DEAD
    else
        rand() <= rule.prob_regrowth ? ALIVE : DEAD
    end

init = fill(ALIVE, 400, 400)
ruleset = Ruleset(ForestFire(); init=init)

output = GtkOutput(init; fps=25, minval=DEAD, maxval=BURNING,
                   processor=ColorProcessor(scheme=ColorSchemes.rainbow, zerocolor=RGB24(0.0)))

sim!(output, ruleset; tspan=(1, 200))
savegif("forestfire_fixed.gif", output)
```
![forestfire_fixed](https://user-images.githubusercontent.com/2534009/72052469-5450c580-3319-11ea-8948-5196d1c6fd33.gif)

And it's a little slower with the extra `rand()` and no zeros optimisation, but not too bad.
```
output = ArrayOutput(init, 50)
@time sim!(output, ruleset; tspan=(1, 50))
  0.318592 seconds (190 allocations: 2.530 MiB, 1.79% gc time)
```

It would be good to see the performance comparison with Agents.jl

**Comment by Datseris** [comment]
Thanks a lot. Me or @kavir1698 will post a benchmark soon. Just so we also have a fair comparison, what are the specs of the PC that run this?

**Comment by rafaqz** [comment]
It's a Lenovo t430. But if you post the code I can run it, or you can run my code above

**Comment by Datseris** [comment]
Yeah, I don't think we can even compare the speeds for this kind of an application. In agents this means we have to create a dictionary of `400*400` agents, while for you this simply means a 400x400 matrix of Integers. Simply by the datastructure you know that there will be massive differences. 

For me, just _initializing_ the model with the same parameters as you takes several minutes, as I have to create a dictionary and add 400x400 agents at each point of the graph.
```
time to initialize model
  126.915472 seconds (4.78 M allocations: 188.208 MiB, 0.06% gc time)
time to do a single step!
  0.509629 seconds (406 allocations: 1.233 MiB)
```
You see that 1 step in Agents takes as much as 50 here (actually I am honestly surprised how fast Agents is w.r.t. DynamicGrids.jl; I expected several orders of magnitude difference due to our internal structure).

Yet another reason that in such applications Agents.jl will always be slower, is because each Agent must have a unique identity. This means that when a tree dies, the "agent" must be literally removed from memory and a brand new one to be created. (this is on purpose and not pointless for more advanced agent based modelling).

In https://github.com/JuliaDynamics/Agents.jl/pull/127 I add this benchmark files, and make a quite big case in Agents.jl on when one should NOT use it, and go with DynamicGrids.jl instead.

I think it would be fair that you add similar comments to your docs, highlighting the target of DynamicGrids.jl and instructing users on when they should use something else. Or, if you are thinking of going forwards with #11 , I can do it.

**Comment by Datseris** [comment]
(see the file `docs/src/index.md`)

**Comment by rafaqz** [comment]
Ok thanks. I think you would have to run all 50 to really compare... it will probably be more than 2 orders of magnitude if that is the architecture. You will have to delete far more agents when they are dying, and that hasn't happened yet at the first step.

A few minor details with your comparison: we do use it for movement of agents - the key difference is that individual agents aren't tracked in their movement between cells, not that movement doesn't occur - we really built it to model movement. We also have much wider range of models than cellular automata - such as gravity model dispersal (jumps between cells based on some gravity factor like human population) and other kinds of random jump models, not to mention interaction models between multiple grids.

**Comment by Datseris** [comment]
Well, can you make the job easier for me by telling me how these bullet points should be instead? I am only an expert of Agents.jl not of DynamicGrids.jl :)

The speed difference I cite is true, the average step takes from 0.3 to 0.8. I always compile on the first step anyway...

**Comment by rafaqz** [comment]
It is just a smidge on the high side of 2 orders ;)

```julia
julia> @time for i in 1:50
           step!(forest, dummystep, forest_step!)
       end
 31.996113 seconds (108.01 k allocations: 64.332 MiB)
```

**Comment by rafaqz** [comment]
>* Is my model a cellular-automaton-like?
>* My agents do not move around.
>* A node will always have a single "agent"/property whose value will change over time.
>* The direct output of the simulation (e.g. the `Array` of cell values over time) is enough.

I think this might be more accurate and to the point:
* Does my model require tracking of individual agents throughout the simulation?

If yes, use Agents.jl. If no, use DynamicGrids.jl

Mostly because in DYnamicGrids.jl:
- cells can have multiple agents, size classes etc if you want to do that, we are working on multi-genome dispersal models, for example.
- Individual can move around, but the movement is'nt traceable afterwards. 
- Cellula automata models are just part the `NeighborhoodRule` components of DynamicGrids.jl, but there are `CellRule`, `PartialRule`, and `Interaction` etc that do other things. 

I've added something like the line above to the DynamicGrids.jl readme, along with the forest fire example.

**Comment by Datseris** [comment]
I don't think what you have written is fair and this is certainly not enough to distinguish the two:

> * Does my model require tracking of individual agents throughout the simulation?

# 1

In DynamicGrids.jl is it possible to make models on arbitrary graphs? In my understanding models can only be run on rectangular grids. 

# 2

In addition in Agents.jl we have done a ton of work about data collection and data processing, which as far as I can see it does not exist here either. For example you can keep which properties of the agents to save as an ouput, but at the moment at least `ArrayOutput` will save everything (which would be Vector of Arrays of Tuples). Agents.jl integrates with DataFrames which makes further data processing easier. 

If my simulation would be done in less than 10 minutes or so, I could easily imagine myself going with the "slow" Agents.jl if it would make my subsequent scientific work so much easier.

# 3
Another important aspect of data processing _during the simulation_ is the aggregation: Agents.jl allows you to not even store the output, but immediatelly process it into an aggregated value like its average or any custom aggregating function. I can imagine a scenario of memory problems if you want to run large models, with multiple properties at each cell, for many time units. You simply wouldn't have the memory to store the entire output, but the aggregated output is easy to store.

# 4

I have to point out that 

> Individual can move around

is misleading phrasing. To phrase it more accurately, the individual in DynamicGrids.jl is the grid cell, and this does not move around. The property of the cells features _dynamic movement_ instead. In your scientific scenario this is the same as "individuals can move around" but in mine it isn't.

# 5

We also have not tested performance with respect to complicated numerical values e.g. `Tuple` on the grid. In general Agents have multiple properties, and rarely have only one. Sometimes you need multiple properties to store previous values, as you cannot update "cells" or agents sequentially as in this [opinion dynamics model](https://juliadynamics.github.io/Agents.jl/dev/examples/HK/). The performance of Agents.jl would be mostly unaffected in this scenario, as the big hit comes from the core datastructure, and increasing the number of the fields of the agents would have only mild consequences and mostly on the model initialization.

However going from `Matrix{Int}` to `Matrix{Tuple{Float64, Floa64}}` is clearly a quite different datastructure. For me it is not obvious that the performance would remain at two orders of magnitude in such a case.

# 6

You claim that you can do 

> live simulation and interaction

in DynamicGrids.jl but not in Agents,jl? What does this even mean? You can't change the rules of your model interactively, you can only change the entire model. Same as in Agents.jl. You can step your model step-by-step interactively, same as in Agents.jl. You can change the numeric value of each cell interactively, same as in Agents.jl.

So...?

---

In conclusion, just as users of Agents.jl can benefit from DynamicGrids.jl for performance gain in specific scenarios, so can users of DynamicGrids.jl benefit from Agents.jl for many other reasons regarding e.g. data collection or making more complex models. You latest README seems to specifically target and discuss only performance, which reduces the entire comparison to "DynamicGrids.jl is two orders faster", but come on, we are scientist. We cannot present such a small part of reality and call it a day.

**Comment by rafaqz** [comment]
Ok we should probably relax the tone a little here. The performance of this package is the results of weeks of my work. I make a point that it is 2 orders of magnitude, not 1 or 2, because that is the actual benchmark, and because I worked _very_ hard to get it there.

Anyway, to answer these points.

1. This package is called DynamicGrids.jl so, _gridded_ data is the domain where Agents.jl is a similar package. It doesn't do graphs or networks and makes no claims to, so I hadn't even thought of mentioning that. You absolutely should, I agree.

2. Actually in dynamic grids `Output`s are modular and easy to extend. Our specific packages normally use [custom outputs](https://github.com/cesaraustralia/Dispersal.jl/blob/master/src/optimisation/objectives.jl#L63), say to sum the output for a loss function on the fly to save the L2 cache for the simulation grids. This one is a little complex as it's highly optimised, only copying blocks that it have things in them, for example. But they don't have to be. I wrote this one today, it just sums the whole `:cost` grid in a multi-grid simulations:

```julia
@default_kw mutable struct CostOutput{T} <: Output{T}
    frames::T        | []
    running::Bool    | false
    starttime::Any   | 1
    stoptime::Any    | 1
    gridname::Symbol | :cost
end
CostOutput(length::Integer; kwargs...) =
    CostOutput(; frames=zeros(length), kwargs...)

DynamicGrids.storeframe!(output::CostOutput, data::MultiSimData, f) =
     output[f] = sum(data[output.gridname])
```

3. As above. Outputs are pretty flexible, you can do whatever processing you want during the simulation and store it however you want on the fly. You can also do step by step simulations, but don't really need to as outputs deal with that.

4. This is really just differences in how we are defining terms. Our simulations are often for modelling organism _movement_. We simulate the movements of groups and individual organisms moving from one cell to another. To us the cell is the location that holds the changing population, and neighborhood dispersal and random jumps are movements of organisms. So tomato/tomaato.

5. DynamicGrids.jl can hold anything in the grid. A tuple of Float32 is bitstype and identical in memory to a single Float64 - so we should expect it to perform identically. The same goes for structs or StaticArrays. Larger objects will be a little slower, but we will all have that problem. Lets not speculate on hypothetical benchmarks.

6. You _can_ change the rules interactively! DynamicGridsInteract.jl does live simulation with real-time parameter sliders for all model parameters, even for arbitrary composite models you chain together. The whole composite model is rebuilt between frames as the simulation runs, and you tweak the sliders. It's basically like a video game for science lol. It even works over the internet. This is also one of the coolest things about the packages that really nothing else does, especially not automated like this. I spent a lot of time writing Flatten.jl and FieldMetadata.jl for this to be possible. You can use them in Agents.jl if you want to.

So

Dynamic grids definitely doesn't track specific individuals movement or specific details about individuals, it's not for that. And it doesn't do graphs or other structures - it just does grids where the individuals are the identical, or are grouped in categories.

But it is a better in many ways at these gridded simulations, not just for the 100x speedup. The live simulations are unique, and composing models by chaining them together is powerful and flexible.
Using multiple grids with interactions between them means you can put together very complicated models from simple components. There are a lot of reasons to prefer it, although of course there are some things Agents.jl may do better.

**Comment by Datseris** [comment]
Sure you are right, I didn't want to have an aggressive tone (although many say I always come out like this, life is tough I guess). And I don't know why you would mention the effort you've put into this package: it's obvious that its a good package that took a lot of effort, I don't think anyone doubted that. I already believed so before knowing about the data processing capabilities, so now I just believe doubly so. If I didn't, I wouldn't be spending a _second_ trying to instruct users of Agents.jl, the package _I_ have invested time, to start using _your_ package instead (honestly, think about it).

---

I think you should comment/explain more your point 6, because from this definition:
```
struct ForestFire{R,N,PC,PR} <: NeighborhoodRule{R}
    neighborhood::N
    prob_combustion::PC
    prob_regrowth::PR
end

@inline DynamicGrids.applyrule(rule::ForestFire, data, state::Integer, index, hood_buffer) =
    if state == ALIVE
        if BURNING in hood_buffer
            BURNING
        else
            rand() <= rule.prob_combustion ? BURNING : ALIVE
        end
    elseif state in BURNING
        DEAD
    else
        rand() <= rule.prob_regrowth ? ALIVE : DEAD
    end

```
I can see that the "rules" of the game are engraved in the model as _type parameters_ (in Julia every function is its own type), which to the best of my knowledge cannot be changed unless initializing a new struct. Are you referring to changing the _values of the parameters of the model_ interactively? Because this can be done anywhere, it is the process of changing the _value_ of the field of a mutable struct. As I have presented it and understood it so far, I don't think this is something as innovative as you present it (and of course you could do it in Agents.jl or any Julia package that does simulations like DifferentialEquations.jl), and so I am probably missing something. You should explain this more... Like, how would you increase the neighborhood size to 2 while the simulation is running? `1` in the above example is a type parameter. (by the way in Agents.jl it is possible to increase the size of the neighborhood both interactively as well as programmatically while the model runs, I don't know if this is a different or not).

The fundamental design core between Agents and DynamicGrids is so much different that performance will never be comparable, DynamicGrids will always be faster for the same problems. But of course this decision allows for different class of complexity in Agents. Regardless, because you _can_ in fact make similar models in Agents and DynamicGrids, users should be able to make an educated choice of what to use.

After this long discussion, here is the most condensed way to present the info on how to decide, which in my eyes is also the most fair:

> Use DynamicGrids.jl and the derivative packages like Dispersion.jl when your model lives on a rectangular grid, while the value of each grid cell does not have an identity (equivalently, the "identity" of each entity in your model is equivalent with its grid cell). Use Agents.jl when you want to (a) use individual agents whose identity (and other properties) is detached from their location (b) multiple agents can occupy the same location or (c) the spatial model of your structure is an arbitrary graph instead of a rectangular grid.

Since I can now see (thank you) that similar dataprocessing is possible here, and since in DynamicGrids is much faster than Agents in its simulations, users should _always_ use DynamicGrids.jl in these conditions (which as far as I can tell, are the only differing points). The point that could be phrased better is the one regarding "identities", let me know how to phrase it better.

I am not sure whether I should mention (b) as a difference or not. In principle you could also have more "agents" in the grid cells of DynamicGrids.jl by making the grid value be a `Vector` of whatever values (or Tuples of values) the agents represent. What I am not sure about is (1) how this would affect your performance and optimizations or (2) the complexity of setting up the model (as far as Agents.jl performance is concerned, it only matters how many agents you have not where they are located)

---

I am very happy to see that the true core of our packages is so different. This means that there isn't much overlap, which is always great, and it also means that users can have faster alternatives for such systems. I am honestly curious why this package has so few stars.

---

Can you paste a gif of this slider business you are talking about? seems very interesting... Are you using Interact.jl under the hood? The plotting capabilities of this package are extremely strong, I wish we could have something like that in Agents.jl. Alas, what makes this difficult (and also why I do not mention plotting in our comparison quote), is that this simply doesn't work for arbitrary graphs :(

**Comment by rafaqz** [comment]
Yes it uses Interact.jl for sliders. But Flatten.jl is the component that translates model structs and tuples into parameters for sliders. 

Say the game of life has a tuple of values like s = (2, 3) for empty cells and b = (2,) for full cells (or whatever it is). You will get three sliders that can change those numbers live. So you can swap the game of life mode as it runs while watching it live. Maybe other things do that.

The thing that is radical is that it is generic - there is no game of life code to make this work. A model you make yourself will also get sliders, and 10 rules chained together will too. You just load the InteractOutput with the ruleset and you have a live controllable interface. You do need a little metadata on some structs to set slider bounds and remove unwanted parameters, but it's fairly straightforward.

I'm nearly always using multiple rules chained together for real simulations. So hooking up a new interface manually every time would not be possible, especially at the rate we experiment with model combinations.

I have been meaning to add a gif to DynamicGridsInteract.jl for ages but I haven't had time yet, I'll try to over the weekend. 


The case where you have multiple different tracked agents in the cell is really best for Agents.jl. It's just hard to do in a grid. When we have multiple agents in a cell it's just storing a 2 instead of a 1, or (1, 2, 3) instead of (2, 3, 4) in a case where you have 3 categories of individuals. You could push agents to vectors for each cell, which could work, but would lose a lot of the performance and would need some custom visualisation code (which is also modular so not that hard).

This is really the case I'm trying to say Agents.jl is better for - where the things need to be tracked and/or are different to all the other things :) 

This package has so few stars because I haven't even announced it anywhere. It was only registered to make Dispersal.jl development easier. Our ecological work definitely comes before promoting this as a general package, and it's broader use is just a side effect we are trying to facilitate, not the main purpose. It's also still changes a lot, so I don't want people complaining about that, and I have 15 other packages in progress distracting me, lol.

**Comment by rafaqz** [comment]
To answer the question I missed - you could actually change the type parameters as this is funcitonal-the model is rebuilt, not updated - they are immutable. 

But we don't, because flatten doesn't work like that - it only looks at fields. I chose for the neighborhood to be a type parameter for performance, but it could actually be a parameter in `Life()` as I don't think the performance is actually impacted. If it was a field, you would get a slider for it, and could change it live.

**Comment by Datseris** [comment]
Okay, so then should we go forwards with the existing quoted text in the readme/documentation: 

> Use DynamicGrids.jl and the derivative packages like Dispersion.jl when your model lives on a rectangular grid, while the value of each grid cell does not have an identity (equivalently, the "identity" of each entity in your model is equivalent with its grid cell). Use Agents.jl when you want to (a) use individual agents whose identity (and other properties) is detached from their location (b) multiple agents can occupy the same location or (c) the spatial model of your structure is an arbitrary graph instead of a rectangular grid.

?

---

> as I don't think the performance is actually impacted

If it isn't, making it a field instead of type parameter is much more clear documentation-wise, as far as newcomers are concerned.

---

> Our ecological work definitely comes before promoting this as a general package, and it's broader use is just a side effect we are trying to facilitate, not the main purpose. 

I can see your point(s), but you should also consider the possibility that getting more popularity can result in getting some quality contributions. It's a risk, as you can't know for sure whether you will get them, but if you do get them you have a huge net gain w.r.t. the time you've spent into getting this popularity. Things like #11 also help with that.

**Comment by rafaqz** [comment]
That seems like a good description - identity is a good word to explain the difference. (Also it's actually "Dispersal.jl")

**Comment by Datseris** [comment]
Okay, updated https://github.com/JuliaDynamics/Agents.jl/pull/127 

*sorry, I used to be a solid state physicist... (told you people can confuse it!)*

**Comment by rafaqz** [comment]
Just fyi, (and to keep the gifs in this issue). Neighborhoods are now modular iterators! So you can swap them out for this example. I'll update the docs with this.

```
windyhood = CustomNeighborhood((1,1), (1,2), (1,3), (2,1), (3,1))
ruleset = Ruleset(ForestFire(; neighborhood=windyhood); init=init)
sim!(output, ruleset; tspan=(1, 200))
savegif("windy_forestfire.gif", output)
```
![windy_forestfire](https://user-images.githubusercontent.com/2534009/72198637-a95d1a80-3484-11ea-8b77-25a4a94b3943.gif)

**Comment by Datseris** [comment]
Heh, having a "windy" forest is really amazing to look at
